jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java

```diff
@@ -190,633 +89,348 @@ public Interceptor getInterceptor()
         {
-            synchronized (_inputQ)
+            _semaphore.drainPermits();
+            int read = _contentProducer.read(this::produceRawContent, b, off, len);
+            if (LOG.isDebugEnabled())
+                LOG.debug("read produced {} byte(s)", read);
+            if (read > 0)
+                return read;
+
+            if (LOG.isDebugEnabled())
+                LOG.debug("read error = " + _error);
+            if (_error != null)
+                throw new IOException(_error);
+
+            if (LOG.isDebugEnabled())
+                LOG.debug("read EOF = {}", _eof);
+            if (_eof.isEarly())
+                throw new EofException("Early EOF");
+
+            if (LOG.isDebugEnabled())
+                LOG.debug("read async = {}", isAsync());
+            if (!isAsync())
             {
-                if (_listener != null)
-                    throw new IllegalStateException("ReadListener already set");
-
-                _listener = Objects.requireNonNull(readListener);
-
-                if (isError())
+                if (_eof.isEof())
+                {
+                    _eof = Eof.CONSUMED_EOF;
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("read on EOF, switching to CONSUMED_EOF and returning");
+                    return -1;
+                }
+                if (LOG.isDebugEnabled())
+                    LOG.debug("read blocked");
+                blockForContent();
+                if (LOG.isDebugEnabled())
+                    LOG.debug("read unblocked");
+            }
+            else
+            {
+                if (_eof.isEof())
                 {
-                    woken = _channelState.onReadReady();
+                    boolean wasInAsyncWait = _channelState.onReadEof();
+                    if (wasInAsyncWait)
+                        scheduleReadListenerNotification();
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("async read on EOF (was in async wait? {}), switching to CONSUMED_EOF and returning", wasInAsyncWait);
+                    _eof = Eof.CONSUMED_EOF;
+                    return -1;
                 }
                 else
                 {
-                    Content content = produceNextContext();
-                    if (content != null)
-                    {
-                        _state = ASYNC;
-                        woken = _channelState.onReadReady();
-                    }
-                    else if (_state == EOF)
-                    {
-                        _state = AEOF;
-                        woken = _channelState.onReadEof();
-                    }
-                    else
-                    {
-                        _state = ASYNC;
-                        _channelState.onReadUnready();
-                        _waitingForContent = true;
-                    }
+                    //TODO returning 0 breaks the InputStream contract. Shouldn't IOException be thrown instead?
```

Comment:
As per https://github.com/eclipse-ee4j/servlet-api/issues/301 throwing an ISE here is good.
We just have to make sure that we never call OnDataAvailable because of raw content that get's entirely consumed by interceptors

Reply:
I saw that, but I'd prefer to keep this the way it is, and fix all the missing ISE in a different PR.

Reply:
ack!